lib/experimental/cmdline.nim

```diff
@@ -0,0 +1,2859 @@
+    else:
+      "-" & name
+
+  func formatFlag(cli: Cli[T], flag: FlagId, name: string): string =
+    let
+      placeholder = cli.placeholderOf(flag)
+      suffix =
+        if not cli.isValueOptional(flag):
+          if placeholder != "":
+            " <" & placeholder & ">"
+          else:
+            " <VALUE>"
+        else:
+          if placeholder == "":
+            ""
+          else:
+            "[=<" & placeholder & ">]"
+
+    prefixedFlag(name) & suffix
+
+  if error of UnknownFlagError:
+    let error = (ref UnknownFlagError)(error)
+    "unexpected flag '" & prefixedFlag(error.flagName) & "' found"
+  elif error of MissingValueError:
+    let error = (ref MissingValueError)(error)
+    "a value is required for '" & cli.formatFlag(error.flag, error.flagName) & "' but none was supplied"
+  elif error of InvalidValueError:
+    let
+      error = (ref InvalidValueError)(error)
+      msg =
+        if error.parent != nil:
+          ": " & error.parent.msg
+        else:
+          ""
+    "invalid value for '" & cli.formatFlag(error.flag, error.flagName) & "'" & msg
+  elif error of FlagError:
+    let error = (ref FlagError)(error)
+    "could not parse flag '" & prefixedFlag(error.flagName) & "': " & error.msg
+  elif error of UnknownPositionalError:
+    let error = (ref UnknownPositionalError)(error)
+    "unexpected positional argument '" & error.positionalValue & "' found"
+  elif error of MissingPositionalError:
+    let error = (ref MissingPositionalError)(error)
+    "missing required positional argument: " & cli.displayOf(error.positional)
+  elif error of InvalidPositionalError:
+    let
+      error = (ref InvalidPositionalError)(error)
+      msg =
+        if error.parent != nil:
+          ": " & error.parent.msg
+        else:
+          ""
+    "invalid value for '" & cli.displayOf(error.positional) & "'" & msg
+  elif error of PositionalError:
+    let error = (ref PositionalError)(error)
+    "could not parse positional parameter '" & error.positionalValue & "': " & error.msg
+  elif error of UnknownCommandError:
+    let error = (ref UnknownCommandError)(error)
+    "unrecognized subcommand: " & error.commandName
+  elif error of MissingCommandError:
+    "missing subcommand"
+  elif error of InvalidCommandError:
+    let
+      error = (ref InvalidCommandError)(error)
+      msg =
+        if error.parent != nil:
+          ": " & error.parent.msg
+        else:
+          ""
+    "invalid subcommand '" & error.commandName & "'" & msg
+  elif error of CommandError:
+    let error = (ref CommandError)(error)
+    "could not parse subcommand '" & error.commandName & "': " & error.msg
+  elif error of HelpError:
+    "help requested"
+  else:
+    "unexpected error parsing command line: " & error.msg
+
+# FIXME: Change the detection to stdmsg once
+# https://github.com/nim-works/nimskull/pull/1547 is merged
+when declared(os.commandLineParams) and declared(stderr):
```

Comment:
Why not move these procedures into a separate module?

Reply:
The UX is horrible, since users would have to go import an another module to actually use cmdline (for typical usage).

If I use `export`, then the exported module is "invisible" in the generated docs, not to mention that it wouldn't simplify any of these conditionals.

Reply:
> If I use `export`, then the exported module is "invisible" in the generated docs

What do you mean be invisible?

> not to mention that it wouldn't simplify any of these conditionals.

Why? There'd be no need for any conditional when `run` is in a separate module. If one imports the separate module with the VM/JS backend, compilation would simply fail.

Reply:
> What do you mean be invisible?

The docs won't show up. It's a tradeoff I'd take for something like `cmdline/parsers`, since they're mostly helpers that you wouldn't normally interact with, not for something crucial to the typical usage flow.

> Why? There'd be no need for any conditional when `run` is in a separate module. If one imports the separate module with the VM/JS backend, compilation would simply fail.

I consider requiring to import two modules a deal breaker for cmdline, where the typical flow ends with user calling `run`.

Reply:
> I consider requiring to import two modules a deal breaker for cmdline, where the typical flow ends with user calling `run`.

You wouldn't need to import two modules, just the one providing `run`, as that also exports `cmdline`. The routines not being listed directly on the separate module's documentation page can be easily helped with by linking to `cmdline` at the top, in my opinion.

I'd prefer there being no `when` workarounds here, given the current state of multi-platform/backend support, but in the end, I don't consider the presence of `when` a blocker for this PR.